plugins/woocommerce/src/Internal/PushNotifications/DataStores/PushTokensDataStore.php

```diff
@@ -169,25 +149,172 @@ public function update( &$push_token ) {
 	}
 
 	/**
-	 * Returns an array of post meta objects as key => value pairs.
+	 * Deletes a push token.
+	 *
+	 * @since 10.5.0
+	 * @param PushToken $push_token An instance of PushToken.
+	 * @throws InvalidArgumentException If the token can't be deleted.
+	 * @throws PushTokenNotFoundException If the token can't be found.
+	 * @return bool
+	 */
+	public function delete( PushToken &$push_token ): bool {
+		if ( ! $push_token->can_be_deleted() ) {
+			throw new InvalidArgumentException(
+				'Can\'t delete push token because the push token data provided is invalid.'
+			);
+		}
+
+		$post = get_post( $push_token->get_id() );
+
+		if ( ! $post || PushToken::POST_TYPE !== $post->post_type ) {
+			throw new PushTokenNotFoundException( 'Push token could not be found.' );
+		}
+
+		wp_delete_post( (int) $push_token->get_id(), true );
+
+		return true;
+	}
+
+	/**
+	 * Find tokens for this user and platform that match either the token
+	 * or device UUID. We check the token value to avoid creating a duplicate.
+	 * We check the device UUID value because only one token should be issued
+	 * per device, therefore if we already have one then we can update it to
+	 * avoid creating a duplicate.
+	 *
+	 * @since 10.5.0
+	 * @param PushToken $push_token An instance of PushToken.
+	 * @return null|PushToken
+	 * @throws InvalidArgumentException If push token is missing data.
+	 */
+	public function get_by_token_or_device_id( PushToken &$push_token ): ?PushToken {
+		if (
+			! $push_token->get_user_id()
+			|| ! $push_token->get_platform()
+			|| ! $push_token->get_origin()
+			|| (
+				/**
+				 * Platforms iOS and Android require token OR device UUID.
+				 */
+				$push_token->get_platform() !== PushToken::PLATFORM_BROWSER
+				&& ! $push_token->get_token()
+				&& ! $push_token->get_device_uuid()
+			)
+			|| (
+				/**
+				 * Browsers don't have device UUIDs, so require token.
+				 */
+				$push_token->get_platform() === PushToken::PLATFORM_BROWSER
+				&& ! $push_token->get_token()
+			)
+		) {
+			throw new InvalidArgumentException(
+				'Can\'t retrieve push token because the push token data provided is invalid.'
+			);
+		}
+
+		global $wpdb;
+
+		$posts = $wpdb->get_results(
+			$wpdb->prepare(
+				"SELECT *
+				FROM {$wpdb->posts}
+				WHERE post_type = %s
+					AND post_status = 'private'
+					AND post_author = %d
+				ORDER BY ID DESC",
+				PushToken::POST_TYPE,
+				$push_token->get_user_id()
+			)
+		);
+
+		foreach ( $posts as $post ) {
+			$candidate = new PushToken();
+			$candidate->set_id( (int) $post->ID );
+
```

Comment:
Even if the expected number of tokens here is small, I believe it'd be much faster to perform another `WP_Query`  here if necessary. You can build the meta part using `meta_query` and use `post__in` to limit the query to the IDs found with the first query.

Let me know if I'm missing something about the `$meta` array that makes it not suitable for `meta_query`!

Reply:
I don't know if I'm understanding correctly - you [previously commented](https://github.com/woocommerce/woocommerce/pull/62024#discussion_r2547428947) that doing a join for each meta item (as the original query did) would be rough performance-wise, but using `WP_Query` to query posts with the post IDs from the first query and filtering results using `meta_query` appears to do the same thing - if I dump the query being performed by `WP_Query`, there's a join for each piece of meta.

It seems like the most performant approach might be to do a query with `$wpdb` to get all meta for the post IDs returned in the first query (and then use that to filter the tokens in PHP) but I can't find a way to achieve that with `WP_Query`. Let me know if I'm on the wrong track here 😕

Reply:
That makes sense, I've added the call to `update_meta_cache` - thanks for helping me understand how `update_meta_cache` works!